### ISSUE TITLE
Python GRPC Memory Leak

### ISSUE THREAD

**Issue by GuanLuo** [grpc/grpc#28513]
### What version of gRPC and what language are you using?
1.43.0 / Python

### What operating system (Linux, Windows,...) and version?
Linux

### What runtime / compiler are you using (e.g. python version or version of gcc)
3.8.10

### What did you do?
Please provide either 1) A unit test for reproducing the bug or 2) Specific steps for us to follow to reproduce the bug. If there’s not enough information to debug the problem, gRPC team may close the issue at their discretion. You’re welcome to re-open the issue once you have a reproduction.

We have a client library that builds on top of GRPC and have test for unbounded memory growth (which repeatedly create client / send request / clean up client [detail](https://github.com/triton-inference-server/client/blob/main/src/python/examples/memory_growth_test.py#L65-L98)). No memory leak was shown in previous GRPC version until we picked up 1.43.0

### What did you expect to see?

Massif report shows flat line indicating the memory usage doesn't increase over time.

### What did you see instead?

Massif report shows increasing memory usage over time.

The report shows that at the end of the test, the largest memory usage is from `cygrpc.cpython`:
```
98.37% (23,549,122B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->56.57% (13,543,424B) 0x9748EFC: void std::deque<unsigned long, std::allocator<unsigned long> >::_M_push_back_aux<unsigned long const&>(unsigned long const&) (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
| ->56.57% (13,543,424B) 0x9746E76: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   ->28.29% (6,771,712B) 0x974715E: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   | ->28.29% (6,771,712B) 0x97474FB: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |   ->28.29% (6,771,712B) 0x964A2EA: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |     ->28.29% (6,771,712B) 0x964D28F: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |       ->28.29% (6,771,712B) 0x963C1A8: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |         ->28.29% (6,771,712B) 0x9724FA3: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |           ->28.29% (6,771,712B) 0x97252F6: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |             ->28.29% (6,771,712B) 0x972557C: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |               ->28.29% (6,771,712B) 0x9713728: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|   |                 ->28.29% (6,771,712B) 0x4A53608: start_thread (pthread_create.c:477)
|   |                   ->28.29% (6,771,712B) 0x497A292: clone (clone.S:95)
|   |
|   ->28.29% (6,771,712B) 0x9747BC4: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     ->14.14% (3,385,856B) 0x9644733: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     | ->14.14% (3,385,856B) 0x964BB24: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     |   ->14.14% (3,385,856B) 0x964D28F: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     |     ->14.14% (3,385,856B) 0x963C1A8: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     |       ->14.14% (3,385,856B) 0x9724FA3: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     |         ->14.14% (3,385,856B) 0x97252F6: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     |           ->14.14% (3,385,856B) 0x972557C: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     |             ->14.14% (3,385,856B) 0x9713728: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|     |               ->14.14% (3,385,856B) 0x4A53608: start_thread (pthread_create.c:477)
|     |                 ->14.14% (3,385,856B) 0x497A292: clone (clone.S:95)
|     |
|     ->14.14% (3,385,856B) 0x9644B03: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|       ->14.14% (3,385,856B) 0x964F5F7: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|         ->14.14% (3,385,856B) 0x9650982: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|           ->14.14% (3,385,856B) 0x9718C00: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|             ->14.14% (3,385,856B) 0x9724F5E: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|               ->14.13% (3,381,760B) 0x97252F6: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|               | ->14.13% (3,381,760B) 0x972557C: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|               |   ->14.13% (3,381,760B) 0x9713728: ??? (in /usr/local/lib/python3.8/dist-packages/grpc/_cython/cygrpc.cpython-38-x86_64-linux-gnu.so)
|               |     ->14.13% (3,381,760B) 0x4A53608: start_thread (pthread_create.c:477)
|               |       ->14.13% (3,381,760B) 0x497A292: clone (clone.S:95)
|               |
|               ->00.02% (4,096B) in 1+ places, all below ms_print's threshold (01.00%)
......
```

**Comment by gnossen** [comment]
Thank you for the reproduction case! I will try to reproduce and get back to you with the results.

**Comment by GuanLuo** [comment]
Is there any update on the issue?

**Comment by brightsparc** [comment]
Hi, just wondered if this issue had been resolved in a more recent release @gnossen ?

**Comment by gnossen** [comment]
Sorry for the long delay. This totally fell off my radar.

@GuanLuo Is this actually still an issue? I can't reproduce this with any version of gRPC.

This is [your memory profiler](https://github.com/triton-inference-server/client/blob/main/src/python/examples/memory_growth_test.py) for 1.42.0 (the version you're currently pinned to):

```

    MB
7.687^                                                                   :    
     |                       ##:::::::@::::::::::::@@:::@@::::::::@@::::::::::
     |                      @# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |                   :::@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |                  :: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |               ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |             @@::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |            :@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |            :@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |           @:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |           @:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |          @@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |         :@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |        @:@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |       :@:@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |       :@:@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |     :::@:@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |  :::: :@:@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |  : :: :@:@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
     |  : :: :@:@@:@ ::::: :@# :::: : @ ::: ::: : :@ : :@ ::: :: :@ :: :::::::
   0 +----------------------------------------------------------------------->Gi
     0                                                                   1.462
```

This is 1.43.0, the version that you filed this issue for:

```
    MB
7.689^                       ##                                               
     |                       # ::::::::::@@:::::@@:::::::::@::::::::::@@::::@:
     |                      @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |                   :@@@# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |                 @@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |                :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |              :::@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |             :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |             :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |           @@:: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |           @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |         @@@ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |         @ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |        :@ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |      :::@ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |      : :@ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |     :: :@ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     |  ::::: :@ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     | :: ::: :@ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
     | :: ::: :@ @ :: :@@:@ @# :::: :: ::@ :: ::@ ::: :: : @: ::: ::: @ : : @:
   0 +----------------------------------------------------------------------->Gi
     0                                                                   1.464
```

For 1.51.1 (the latest):

```
    MB
7.635^                                                 ::                     
     |                 ##:::::@::::::::::::::::::::::::: ::::::::::::@:::::@::
     |                :# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |              :::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |             ::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |            :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |          :::::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |          : :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |         @: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |         @: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |        @@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |        @@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |       :@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |     :::@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |     : :@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |    :: :@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |  @::: :@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     |  @::: :@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     | :@::: :@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
     | :@::: :@@: :::::# :: ::@: :: : :: :::::::: :::::: :: :::::::::@:::::@::
   0 +----------------------------------------------------------------------->Gi
     0                                                                   1.944
```

This looks like good in all cases. Anything I might be missing here?

**Comment by gnossen** [comment]
Also, for posterity, the steps for running the repro:
- Clone [the server repo](https://github.com/triton-inference-server/server)
- `mkdir qa/L0_client_memory_growth/models/custom_identity_int32/1` (this step wasn't documented anywhere)
- `docker run --rm -p8000:8000 -p8001:8001 -p8002:8002 -v  $(pwd)/qa/L0_client_memory_growth/models:/models nvcr.io/nvidia/tritonserver:21.08-py3 tritonserver --model-repository=/models`
- Install Valgrind (I had to do this from source to get access to 3.20.0, which has support for dwarf 5)
- Clone [the client repo](https://github.com/triton-inference-server/client)
- `pip install tritonclient`
- `pip uninstall grpcio && pip install grpcio==$VERSION_UNDER_TEST`
- `valgrind --tool=massif python3.7 memory_growth_test.py --protocol=grpc --repetitions=10000`
- `ms_print massif.out.PID`

**Comment by jbkyang-nvi** [comment]
Hi @gnossen as https://github.com/triton-inference-server/client/blob/main/src/python/library/requirements/requirements_grpc.txt#L31 commented 1.43.0+ versions do leak. I'm building with 1.51.0 will give you an update when I get it done.

**Comment by jbkyang-nvi** [comment]
Hi, I ran this on my local machine with similar commands. I think you need to run for a bit longer

```
>valgrind --tool=massif python3 memory_growth_test.py --protocol=grpc --repetitions=100000
> ms_print massif.out.292628
--------------------------------------------------------------------------------
Command:            python3 memory_growth_test.py --protocol=grpc --repetitions=100000
Massif arguments:   (none)
ms_print arguments: massif.out.292628
--------------------------------------------------------------------------------


    MB
10.85^                                                                       #
     |                                                           ::::::@@:@::#
     |                                                ::::::::::::: :: @ :@::#
     |                                    ::::::::@::::: :::: : ::: :: @ :@::#
     |                        ::::::::::::: :: :: @:: :: :::: : ::: :: @ :@::#
     |          @::::::::::::::: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |::::::::::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
     |: : ::: ::@:: : ::::::: :: :: : ::: : :: :: @:: :: :::: : ::: :: @ :@::#
   0 +----------------------------------------------------------------------->Gi
     0                                                                   88.65

Number of snapshots: 55
 Detailed snapshots: [8, 32, 47, 50, 53 (peak)]

```

**Comment by gnossen** [comment]
I also reproduced memory growth at 100000 repetitions, but the memory growth actually seemed to be coming from `json.loads`, not from gRPC:

```
--------------------------------------------------------------------------------
  n        time(i)         total(B)   useful-heap(B) extra-heap(B)    stacks(B)
--------------------------------------------------------------------------------
  0              0                0                0             0            0
  1  2,266,684,803        7,122,064        6,923,167       198,897            0
  2  3,676,900,935        7,161,000        6,961,567       199,433            0
  3  5,719,495,109        7,231,728        7,029,968       201,760            0
  4  8,023,129,370        7,282,432        7,080,871       201,561            0
  5  9,391,981,933        7,325,088        7,122,855       202,233            0
  6 11,329,989,289        7,374,848        7,171,920       202,928            0
  7 13,849,417,808        7,446,320        7,241,639       204,681            0
  8 16,105,328,516        7,559,488        7,355,597       203,891            0
97.30% (7,355,597B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->78.31% (5,919,627B) 0x1B2314: PyMem_RawMalloc (obmalloc.c:520)
| ->78.31% (5,919,627B) 0x1B2314: _PyObject_Malloc (obmalloc.c:1584)
|   ->78.31% (5,919,627B) 0x1B2314: _PyObject_Malloc (obmalloc.c:1576)
|     ->29.41% (2,223,399B) 0x1D7738: PyUnicode_New (unicodeobject.c:1322)
|     | ->23.77% (1,796,730B) 0x1D8D4D: _PyUnicode_FromUCS1 (unicodeobject.c:2216)
|     | | ->23.77% (1,796,730B) 0x241270: r_object (marshal.c:1169)
|     | |   ->23.77% (1,796,730B) 0x241C41: r_object (marshal.c:1231)
|     | |     ->23.77% (1,796,730B) 0x2413B4: r_object (marshal.c:1402)
|     | |     | ->22.30% (1,685,417B) 0x241C41: r_object (marshal.c:1231)
|     | |     | | ->22.30% (1,685,417B) 0x2413B4: r_object (marshal.c:1402)
|     | |     | |   ->18.94% (1,431,918B) 0x2429AA: read_object (marshal.c:1488)
|     | |     | |   | ->18.94% (1,431,918B) 0x2429AA: marshal_loads_impl (marshal.c:1800)
```

[massif.out.2382065.txt](https://github.com/grpc/grpc/files/10282898/massif.out.2382065.txt)


Can you please share your full profile?

**Comment by jbkyang-nvi** [comment]
It looks like 1.51.1 magically fixes the memory growth (although 1.43.0 still leaks). I think we can close this for now (unless you want me to investigate with 1.43.0)

**Comment by XuanWang-Amos** [comment]
Closing this now since issues seems to be resolved, if everyone have this issue again please respond with more logs and reproducing steps.
